substrate/utils/frame/benchmarking-cli/src/pallet/command.rs

```diff
@@ -255,20 +248,75 @@ impl PalletCmd {
 		.with_runtime_cache_size(2)
 		.build();
 
-		let (list, storage_info): (Vec<BenchmarkList>, Vec<StorageInfo>) =
-			Self::exec_state_machine(
-				StateMachine::new(
-					state,
-					&mut changes,
-					&executor,
-					"Benchmark_benchmark_metadata",
-					&(self.extra).encode(),
-					&mut Self::build_extensions(executor.clone()),
-					&runtime_code,
-					CallContext::Offchain,
-				),
-				ERROR_METADATA_NOT_FOUND,
-			)?;
+		let runtime_version: RuntimeVersion = Self::exec_state_machine(
+			StateMachine::new(
+				state,
+				&mut changes,
+				&executor,
+				"Core_version",
+				&[],
+				&mut Self::build_extensions(executor.clone()),
+				&runtime_code,
+				CallContext::Offchain,
+			),
+			ERROR_RUNTIME_VERSION_NOT_FOUND,
+		)?;
+
+		// Get the version of `Benchmark` runtime API.
+		let (_, version) = runtime_version
+			.apis
+			.iter()
+			.find(|(name, _)| name == &blake2_64(b"Benchmark"))
+			.ok_or("Benchmark API not found in runtime version apis")?;
+
+		let RuntimeBenchmarkInfo { list, storage_info, max_extrinsic_weight, db_weight } = {
+			if *version == 1 {
+				// This ensures that we don't run sanity checks on API v1.
+				let old_benchmark_data: (Vec<BenchmarkList>, Vec<StorageInfo>) =
+					Self::exec_state_machine(
+						StateMachine::new(
+							state,
+							&mut changes,
+							&executor,
+							"Benchmark_benchmark_metadata",
+							&(self.extra).encode(),
+							&mut Self::build_extensions(executor.clone()),
+							&runtime_code,
+							CallContext::Offchain,
+						),
+						ERROR_METADATA_NOT_FOUND,
+					)?;
+				old_benchmark_data.into()
```

Comment:
Yes this looks good. So now it should be possible to have old runtimes still work.  
I guess to be really precise, we should keep the old `fn benchmark_metadata` around in the `pub trait Benchmark`, since this is what we call into.  
There is a `changed_in` attribute that you can search for in the code base and add.